### DISCUSSION TITLE
Hardware limits related API design

### DISCUSSION THREAD

**Discussion by gyagp** [gpuweb/gpuweb#2161] [Q&A]
Hardware limits are very important information. Sometimes, we need to know the minimal values supported for portability, while other times, we need the best values (privacy concern can be mitigated so the values don't have to be the best supported ones in real hardware) for more performance. 
Per my understanding, currently in the spec, there are 2 APIs related to this. requestAdapter() will return GPUAdapter, including GPUSupportedLimits. We can get the minimal values here. But to get the best values, we need to use requestDevice() and pass parameter GPUDeviceDescriptor, which includes limits (requiredLimits) we request. If the request couldn't be satisfied, it will return false.
So I have the below questions:
* Why do we need to attach the hardware limits to 2 different concepts, adapter and device? I vaguely remember requestDevice() is more lightweight than requestAdapter().
* Why do we need 2 different ways to get the minimal and best values? The way to get the best values is very inconvenient, as I may have to bisect a range of values before the best value is tested out.

My proposal:
Can we simply call requestAdapter() and both minimal and best values can be returned? Currently we have 26 hardware limits defined in the spec, so I don't think this will bloat a lot. But if this is the concern or we have other concerns (e.g., the queries would take a long time), we may need a new API for query. Anyway, I hope we can have a consistent way to get both values, with a single API.

**Comment by Kangz** [comment]
`requestAdapter` should return the maximum limits supported by the adapter (or at least what the browser will allow you to use). The minimum limits are never returned explicitly in the API (because they are just in the spec) but if a developer when to test against the minimum limits they can call `createDevice` with no extra limits.

**Comment by kainino0x** [comment]
> * Why do we need to attach the hardware limits to 2 different concepts, adapter and device?

The adapter represents the hardware, and the "hardware" limits (or whatever limits the browser exposes) are only attached there. The device limits are independent of the hardware; they're enforced in software. And given the same device limits we enforce the same validation on usage of the device, regardless of what the limits of the underlying adapter (hardware) are.

The benefit of this is that, if a program runs on one GPUDevice, it should run on any machine that can support the same requested limits/features. It reduces the danger of a developer accidentally making a change that requires higher capabilities than the ones they've previously decided to require.
